### PR TITLE
Implement reporting for Flux installation status and statistics

### PR DESCRIPTION
**Pull request by stefanprodan**
**FluxReport** is an API that reflects the observed state of a Flux installation.
Its purpose is to aid in monitoring and troubleshooting Flux by providing information about the installed components and their readiness, the distribution details, reconcilers statistics, cluster sync status, etc.

A single custom resource of this kind can exist in a Kubernetes cluster with the name `flux`. The resource is automatically generated in the same namespace where the flux-operator is deployed and is updated by the operator at regular intervals.

## Example

The following example shows a FluxReport custom resource generated on a cluster where Flux was deployed:

```yaml
apiVersion: fluxcd.controlplane.io/v1
kind: FluxReport
metadata:
  name: flux
  namespace: flux-system
spec:
  components:
    - image: ghcr.io/fluxcd/helm-controller:v1.0.1@sha256:a67a037faa850220ff94d8090253732079589ad9ff10b6ddf294f3b7cd0f3424
      name: helm-controller
      ready: true
      status: 'Current Deployment is available. Replicas: 1'
    - image: ghcr.io/fluxcd/kustomize-controller:v1.3.0@sha256:48a032574dd45c39750ba0f1488e6f1ae36756a38f40976a6b7a588d83acefc1
      name: kustomize-controller
      ready: true
      status: 'Current Deployment is available. Replicas: 1'
    - image: ghcr.io/fluxcd/notification-controller:v1.3.0@sha256:c0fab940c7e578ea519097d36c040238b0cc039ce366fdb753947428bbf0c3d6
      name: notification-controller
      ready: true
      status: 'Current Deployment is available. Replicas: 1'
    - image: ghcr.io/fluxcd/source-controller:v1.3.0@sha256:161da425b16b64dda4b3cec2ba0f8d7442973aba29bb446db3b340626181a0bc
      name: source-controller
      ready: true
      status: 'Current Deployment is available. Replicas: 1'
  distribution:
    entitlement: Issued by controlplane
    managedBy: flux-operator
    status: Installed
    version: v2.3.0
  reconcilers:
    - apiVersion: helm.toolkit.fluxcd.io/v2
      kind: HelmRelease
      stats:
        failing: 1
        running: 42
        suspended: 3
    - apiVersion: kustomize.toolkit.fluxcd.io/v1
      kind: Kustomization
      stats:
        failing: 0
        running: 5
        suspended: 0
    - apiVersion: notification.toolkit.fluxcd.io/v1
      kind: Receiver
      stats:
        failing: 0
        running: 1
        suspended: 0
    - apiVersion: notification.toolkit.fluxcd.io/v1beta3
      kind: Alert
      stats:
        failing: 0
        running: 1
        suspended: 0
    - apiVersion: notification.toolkit.fluxcd.io/v1beta3
      kind: Provider
      stats:
        failing: 0
        running: 1
        suspended: 0
    - apiVersion: source.toolkit.fluxcd.io/v1
      kind: GitRepository
      stats:
        failing: 0
        running: 2
        suspended: 0
        totalSize: 3.7 MiB
    - apiVersion: source.toolkit.fluxcd.io/v1
      kind: HelmChart
      stats:
        failing: 1
        running: 55
        suspended: 0
        totalSize: 15.7 MiB
    - apiVersion: source.toolkit.fluxcd.io/v1
      kind: HelmRepository
      stats:
        failing: 0
        running: 7
        suspended: 3
        totalSize: 40.5 MiB
    - apiVersion: source.toolkit.fluxcd.io/v1beta2
      kind: Bucket
      stats:
        failing: 0
        running: 0
        suspended: 0
    - apiVersion: source.toolkit.fluxcd.io/v1beta2
      kind: OCIRepository
      stats:
        failing: 0
        running: 1
        suspended: 0
        totalSize: 78.1 KiB
  sync:
    id: kustomization/flux-system
    path: clusters/production
    ready: true
    source: https://github.com/my-org/my-fleet.git
    status: 'Applied revision: refs/heads/main@sha1:a90cd1ac35de01c175f7199315d3f4cd60195911'
status:
  conditions:
    - lastTransitionTime: "2024-06-20T19:59:30Z"
      message: Reporting finished in 272ms
      observedGeneration: 4
      reason: ReconciliationSucceeded
      status: "True"
      type: Ready
```

1. Export the report in YAML format:

   ```shell
   kubectl -n flux-system get fluxreport/flux -o yaml
   ```

2. Trigger a reconciliation of the report:

   ```shell
   kubectl -n flux-system annotate --overwrite fluxreport/flux \
    reconcile.fluxcd.io/requestedAt="$(date +%s)"
   ```

3. Change the report reconciliation interval:

   ```shell
   kubectl -n flux-system annotate --overwrite fluxreport/flux \
    fluxcd.controlplane.io/reconcileEvery=5m
   ```

4. Pause the report reconciliation:

   ```shell
   kubectl -n flux-system annotate --overwrite fluxreport/flux \
    fluxcd.controlplane.io/reconcile=disabled
   ```

5. Resume the reconciliation of the report:

   ```shell
    kubectl -n flux-system annotate --overwrite fluxreport/flux \
     fluxcd.controlplane.io/reconcile=enabled
    ```

### CODE REVIEW

File: api/v1/fluxreport_types.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+// FluxComponentStatus defines the observed state of a Flux component.
+type FluxComponentStatus struct {
+	// Name is the name of the Flux component.
+	// +required
+	Name string `json:"name"`
+
+	// Ready is the readiness status of the Flux component.
+	// +required
+	Ready bool `json:"ready"`
```
Comment (souleb) [inline api/v1/fluxreport_types.go:64]: Having `ready` and `status` is confusing I think. Perhaps `ready` and `message`?

File: internal/reporter/reporter.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	if err != nil {
+		return report, fmt.Errorf("failed to compute sync status: %w", err)
+	}
+	report.SyncStatus = syncStatus
+
+	return report, nil
+}
+
+// RequestReportUpdate annotates the FluxReport object to trigger a reconciliation.
+func RequestReportUpdate(ctx context.Context, kubeClient client.Client, instance, manager, namespace string) error {
```
Comment (souleb) [inline internal/reporter/reporter.go:75]:
```suggestion
func RequestReportUpdate(ctx context.Context, kubeClient client.Client, instance, namespace string) error {
```

Comment (stefanprodan) [reply internal/reporter/reporter.go:75]: The manager is needed, I just forgot to set it in the Patch call.

**Review by souleb — APPROVED** [review]
LGTM!

**Pull request merged**